### ISSUE TITLE
Issue writing testes alongside rswag

### ISSUE THREAD

**Issue by Kutomore** [rswag/rswag#217]
Hello, I've been assigned into a project that uses rswag for it's documentation, I have very little experience with it but I've been getting the hang of it. The issue is people weren't actually writing tests, they were just creating the endpoints documentation. So there's no one with a background experience that can help me in my team.

I'm currently using the documentation avaliable in the readme to try and create a simple test.
The run_test! command didn't work due to authentication issues, so I'm trying to do it manually as is explained in the readme.

```
before do
	let!(:product) { create(:product) }
	get '/products', headers: headers
end
it 'returns a valid 200 response with the product' do |example|
	assert_response_matches_metadata(example.metadata)
end
```

Where in the header I handle the authentication part. The only issue is that when I try to run this test I get the following error `  undefined method 'path' for RSpec::ExampleGroups::`

By the nature of the error I assume maybe it's some configurarion missing on Rspec? I couldn't find anything similar on the open issues or in the Readme, any help is appreciated.

**Comment by benbonnet** [comment]
same here :/

is it tied to the rails6 use maybe ?

**Comment by Kutomore** [comment]
@benbonnet probably not, since I'm using rails 5 on this project.

**Comment by jaydorsey** [comment]
Can you post a full test file that's failing? Do you have `require 'swagger_helper'` at the top of your test file?

**Comment by Kutomore** [comment]
So, I hadn't tried it again since the time of writing of this issue, so I decided to try again with the require you said. To my surprise, both using and not using the require gives me the same response, it "works", I mean I don't get the error I was getting before. I have no idea why. But now I still can't get it to work.

```
describe 'Eproducts endpoints' do
	let!(:user) { create(:user) }
	let(:headers) { authenticated_header(user) }
	let!(:product) { create(:eproduct) }

	path '/ecommerce/v2/eproducts' do
		get "Get the product list from the ecommerce" do
			security [ Bearer: [""] ]
			tags 'Ecommerce'
			consumes 'application/json'
			description "Get a list of products from the ecommerce"
			parameter name: :search, :in => :query, :type => :string, :description => 'Search in fields: name, description, sku', required: false
			parameter name: :category_id, :in => :query, :type => :string, :description => 'Search by category id', required: false
			parameter name: :campaign_id, :in => :query, :type => :string, :description => 'Search by category alias', required: false
			parameter name: :price, :in => :query, :type => :integer, :description => 'Search products that are between the past price', required: false
			parameter name: :attributes, :in => :query, :type => :integer, :description => "Search for dynamic attributes, example: attributes=[{alias: 'visible', value: true}, {alias: 'color', value: 'azul}]", required: false
			parameter name: :skus, :in => :query, :type => :integer, :description => "Search for products related to a sku , example: skus=['145aad', 'aaaaaa']", required: false
			parameter name: :campaig_id, :in => :query, :type => :integer, :description => "Search for products related to a campaign, example: campaign_id='145aad-11dd21-asdasd-123123'", required: false
			parameter name: :campaign_alias, :in => :query, :type => :integer, :description => "Search for products related to a campaign, example: campaign_alias='new-campaign'", required: false
			parameter name: :sort, :in => :query, :type => :integer, :description => 'Sort products example: -title, +attributes.color', required: false
			parameter name: :per, :in => :query, :type => :integer, :description => 'results per page', :default => 10, required: false
			parameter name: :page, :in => :query, :type => :integer, :description => 'page', :default => 1, required: false

			response '200', 'Products found' do
				schema type: :array,
							 properties: {
									 data: {
											 type: :array,
											 items: {
													 '$ref': '#/definitions/ecommerce_eproducts_object'
											 }
									 },
									 links: {
											 type: :object,
											 properties: {
													 self: {
															 type: :string,
															 format: :url,
															 example: "//ecommerce/v2/eproducts/?page[number]=1&page[size]=10"
													 },
													 first: {
															 type: :string,
															 format: :url,
															 example: "//ecommerce/v2/eproducts/?page[number]=1&page[size]=10"
													 },
													 prev: {
															 type: :string,
															 format: :url,
															 example: nil
													 },
													 next: {
															 type: :string,
															 format: :url,
															 example: "//ecommerce/v2/eproducts/?page[number]=2&page[size]=10"
													 },
													 last: {
															 type: :string,
															 format: :url,
															 example: "//ecommerce/v2/eproducts/?page[number]=10&page[size]=10"
													 }
											 },
											 required: [ 'self', 'first', 'last']
									 },
									 meta: {
											 '$ref': '#/definitions/meta_pagination_object'
									 }
							 }
				before do
					get '/ecommerce/v2/eproducts', params: { page: 1, per: 1 }, headers: headers
				end
				it 'returns a valid 200 response with the product' do |example|
					assert_response_matches_metadata(example.metadata)
				end
			end

			response '401', 'Unauthorized' do
				let(:id) { 'invalid' }
				run_test!
			end

			response '500', 'Internal Server Error' do
				let(:id) { 'invalid' }
				run_test!
			end
		end
	end
```
this is the test in question. When I attempt to run it at the moment I get the error `Expected response body to match schema: The property '#/' of type object did not match the following type: array in schema 21d9ae67-0c16-5104-a531-ae0d04a887f1`. I used binding.pry to see example.metadata that I'm sending to the matcher, but it seems it's only a "copy" of the documentation I wrote.

**Comment by jaydorsey** [comment]
It looks like your schema doesn't match your response. Can you put a pry in and dump your actual response/result as json? 

Getting the schema syntax to match up w/ actual responses can be a little tricky

**Comment by Kutomore** [comment]
Oh, sorry for the delay, which variable holds these values? The response?

**Comment by jaydorsey** [comment]
> Oh, sorry for the delay, which variable holds these values? The response?

Yes, it should be the `response` object. 

Also, probably unrelated but you could rewrite your before/it assertion block per the docs. There's some magic going on with how metadata gets injected

Instead of this:

	before do
		get '/ecommerce/v2/eproducts', params: { page: 1, per: 1 }, headers: headers
	end
	it 'returns a valid 200 response with the product' do |example|
		assert_response_matches_metadata(example.metadata)
	end

Have you tried this?

    let(:'page') { 1 }
    let(:'per') { 1 }
    let(:'headers') { headers }

    run_test!

The reason i mention it is I'm not sure if manually crafting your request is missing anything. If you want more control over the spec, because `run_test!` is doing too much (it just checks the response code for you, I think), there's an example from the docs:

```ruby
  before do |example|
    submit_request(example.metadata)
  end

  it 'returns a valid 201 response' do |example|
    assert_response_matches_metadata(example.metadata)
  end
```

**Comment by Kutomore** [comment]
So, I'm currently trying to fix this tests as my only task, so again, sorry for the delay but I'm now able to pay full attention to this.

> Have you tried this?
> 
>```
>let(:'page') { 1 }
> let(:'per') { 1 }
> let(:'headers') { headers }
> ```
> run_test!

Yes I did, when I try to do it this way I get an authentication error:
```
Failure/Error: example.run
     
     NoMethodError:
       undefined method `Authorization' for #<RSpec::ExampleGroups::EproductsEndpoints::EcommerceV2Eproducts::Get::ProductsFound:0x0000563912106070>
```
Which I don't get by using the `get` in the before do, since I can pass an authentication Header that I generate in the `headers: headers` option.

Anyway, right now I'm trying to figure out how to make the schema match the response object. But it leads me to believe that my approach of using `get '/ecommerce/v2/eproducts', params: { page: 1, per: 1 }, headers: headers` is also not going to work, because when I call `response.body` inside the `it` block I get:
```
{
  "data": [],
  "links": {
    "self": "http://www.example.com/ecommerce/v2/eproducts?page%5Bnumber%5D=1&page%5Bsize%5D=1&per=1",
    "first": "http://www.example.com/ecommerce/v2/eproducts?page%5Bnumber%5D=1&page%5Bsize%5D=1&per=1",
    "prev": "nil",
    "next": "nil",
    "last": "http://www.example.com/ecommerce/v2/eproducts?page%5Bnumber%5D=1&page%5Bsize%5D=1&per=1"
  },
  "meta": {
    "current_page": 1,
    "next_page": "nil",
    "prev_page": "nil",
    "total_pages": 0,
    "total_count": 0,
    "suggestions": "nil"
  }
}
```
Which tells me there were no Eproducts on the Database when this request was run. Which is odd since inside the pry I can get the `Class.count` or `.first` and see that it's there, even more because I have created the product on the first line of the test, before even the schema. That leads me to believe that when I call `request` inside my pry on the it block it isn't the get request I wrote. Am I right?

**Comment by jaydorsey** [comment]
> Which I don't get by using the get in the before do, since I can pass an authentication Header that I generate in the headers: headers option.

That probably fails because you don't have the securityDefinition correctly in the swagger helper.rb file. It's a little tricky tbh, but there are some good comments in various issues on how to configure it under different scenarios

> That leads me to believe that when I call request inside my pry on the it block it isn't the get request I wrote. 

It would be helpful to see this block, and the before setup (if you changed it) but something like this is how I'd start debugging this:

```
it 'valid response' do
  # This should show you the path that was requested
  puts request.url
  # I'd drop a pry into the controller also; maybe it's your auth headers, 
  # maybe something else is preventing this record from returning in
  # the results?
end
```

What's kind of odd to me is that in your response you have a `per` and a `page[size]` and I'd expect to see one or the other (those are the same thing, right?). Also a `page[number]` but I don't see that in your params either. Not sure if that's updates since your original snippet was shared or something else interfering but wanted to point it out in case it was a bug

**Comment by Kutomore** [comment]
> What's kind of odd to me is that in your response you have a per and a page[size] and I'd expect to see one or the other (those are the same thing, right?). 

Well you just got me there, that is indeed wrong, I'm suposed to send page number and a per for the size, well spotted.

> It would be helpful to see this block, and the before setup (if you changed it)

Since I'm mostly trying to get this to work, my setup is pretty basic, just this:
```
let!(:user) { create(:user) }
let(:headers) { authenticated_header(user) }
let!(:product) { create(:eproduct) }
#the schema and all that stuff
before do
   get '/ecommerce/v2/eproducts', params: { page: 1, per: 1 }, headers: headers
end
it 'returns a valid 200 response with the product' do |example|
  binding.pry
  assert_response_matches_metadata(example.metadata)
end
```

Anyway, I'll try to debugg it like you said.

> That probably fails because you don't have the securityDefinition correctly in the swagger helper.rb file. It's a little tricky tbh, but there are some good comments in various issues on how to configure it under different scenarios

~~Ok, I'll search for it on the issues.~~ Well, you were right, if I do define th Authorization before calling run_test it does work, I mean, I should have figured that one out by myself, sorry, but thanks. I guess all that is left is to make the schema match the response :v

**Comment by jaydorsey** [comment]
> I guess all that is left is to make the schema match the response

If you get stuck and can post the response.body I might be able to help. The syntax for the schema is a little tricky too and I didn't see a lot of complex examples. I will say that I had to struggle with it for a bit (my response was good, but it took me a while to write a good matching schema) but it does get easier once you've knocked a few out

**Comment by Kutomore** [comment]
> If you get stuck and can post the response.body I might be able to help. The syntax for the schema is a little tricky too and I didn't see a lot of complex examples. I will say that I had to struggle with it for a bit (my response was good, but it took me a while to write a good matching schema) but it does get easier once you've knocked a few out

Yeah, you're right once again, I did manage to figure out why it wasn't showing up, turns out one of my filters was acting up since my factory wasn't that well made. My response body is:

```
"{\"data\":[{\"id\":\"ee757c1e-c1c6-4683-8a2a-141669dda3e3\",\"type\":\"ecommerce-eproducts\",\"attributes\":{\"title\":\"Samuel Smith’s Imperial IPA\",\"etype\":\"simple\",\"description\":\"Belgian And French Ale\",\"price\":0.0,\"discounted_price\":0,\"eproduct_product_id\":\"1ea4372a-baf2-4135-baaf-0c3d7e87e31b\",\"stock\":0,\"image\":\"true\",\"sku\":\"828957\"}}],\"links\":{\"self\":\"http://www.example.com/ecommerce/v2/eproducts?page%5Bnumber%5D=1\\u0026page%5Bsize%5D=1\\u0026per=1\",\"first\":\"http://www.example.com/ecommerce/v2/eproducts?page%5Bnumber%5D=1\\u0026page%5Bsize%5D=1\\u0026per=1\",\"prev\":null,\"next\":null,\"last\":\"http://www.example.com/ecommerce/v2/eproducts?page%5Bnumber%5D=1\\u0026page%5Bsize%5D=1\\u0026per=1\"},\"meta\":{\"current_page\":1,\"next_page\":null,\"prev_page\":null,\"total_pages\":1,\"total_count\":1,\"suggestions\":null}}"
```
Or a better formatted version
```
{
  "data": [
    {
      "id": "ee757c1e-c1c6-4683-8a2a-141669dda3e3",
      "type": "ecommerce-eproducts",
      "attributes": {
        "title": "Samuel Smith’s Imperial IPA",
        "etype": "simple",
        "description": "Belgian And French Ale",
        "price": 0,
        "discounted_price": 0,
        "eproduct_product_id": "1ea4372a-baf2-4135-baaf-0c3d7e87e31b",
        "stock": 0,
        "image": "true",
        "sku": "828957"
      }
    }
  ],
  "links": {
    "self": "http://www.example.com/ecommerce/v2/eproducts?page%5Bnumber%5D=1&page%5Bsize%5D=1&per=1",
    "first": "http://www.example.com/ecommerce/v2/eproducts?page%5Bnumber%5D=1&page%5Bsize%5D=1&per=1",
    "prev": null,
    "next": null,
    "last": "http://www.example.com/ecommerce/v2/eproducts?page%5Bnumber%5D=1&page%5Bsize%5D=1&per=1"
  },
  "meta": {
    "current_page": 1,
    "next_page": null,
    "prev_page": null,
    "total_pages": 1,
    "total_count": 1,
    "suggestions": null
  }
}
```
But I still get the same error that I got before:
`Expected response body to match schema: The property '#/' of type object did not match the following type: array in schema 21d9ae67-0c16-5104-a531-ae0d04a887f1`. So I'm currently trying to match all the variables on the schema with the ones on the eproduct creation.

**Comment by jaydorsey** [comment]
It _looks_ OK to me, but I'd also need to see the definition for your `ecommerce_eproducts_object`, can you share that?

A few debugging tricks I use:

- Don't use a reference right now; get it working without the reference first
- If that doesn't work, start with a bare minimum wrapper & keep layering schema stuff on until it matches

Side question: It looks like you're using JSONAPI; what gem are you using? (we're using jsonapi-resources)

I have a generic jsonapi schema wrapper I use to make sure the schema is properly formatted jsonapi that we use on all endpoints, then we check for presence of particular values and keys as needed. It doesn't, for example, validate that you have price, description, etc. or that the "type" key is `ecommerce_eproducts_object` but what I've found is that a lot of the work of defining all those schemas ends up being duplicate work; mostly because we provide an example response with every test that we document (we configured rswag to do that)

I'd _like_ to do bespoke schemas for everything at some point but we haven't had to make it a priority as of yet. We do create/update request schemas though, just not index/show response

Edit: The response schema is [here](https://gist.github.com/jaydorsey/3a5397c24955ec8788828eb364682448)

Source for that is [here](https://github.com/eneuhauser/json-api/blob/schema/schema) but iirc I had to make a change to make it work right with whatever gem I was using, or ruby, or something

There's also some tools for helping you create & maintain schemas [here](https://www.sitepoint.com/document-your-json-api-schema-with-prmd/) (I haven't had time to pursue these in depth yet)

Edit 2: [This post](https://github.com/json-api/json-api/issues/271) might also be helpful, it has some other documentation in there (if you're doing jsonapi)

**Comment by Kutomore** [comment]
> It looks OK to me, but I'd also need to see the definition for your ecommerce_eproducts_object, can you share that?

```
ecommerce_eproducts_object:{
            type: :object,
            properties: {
                id: {
                    type: :string,
                    example: 'fbf10588-6946-4532-af7c-7200dc684e63'
                },
                title: {
                    type: :string,
                    example: 'This product title'
                },
                description:{
                    type: :string,
                    example: 'This product description'
                },
                categories:{
                    type: :array,
                    items: {
			'$ref' => '#/definitions/ecommerce_categories_object'
		  }
                },
                campaigns:{
                    type: :array,
                    items: {
			'$ref' => '#/definitions/ecommerce_campaigns_object'
		  }
                },
                features:{
                    type: :array,
                    items: {
			'$ref' => '#/definitions/ecommerce_features_object'
		    }
                },
                eproducts_products:{
                    type: :array,
                    items: {
			'$ref' => '#/definitions/ecommerce_eproducts_products_object'
		     }
                }
            }
	},
```

> Side question: It looks like you're using JSONAPI; what gem are you using? (we're using jsonapi-resources)

We're currently using `fast_jsonapi`If I'm not mistaken.

>If that doesn't work, start with a bare minimum wrapper & keep layering schema stuff on until it matches

Does the schema have to be an exact match with the response? Like the values and such, since my values are randomly generated with Faker this might be a problem.

>Source for that is here but iirc I had to make a change to make it work right with whatever gem I was using, or ruby, or something
> There's also some tools for helping you create & maintain schemas here (I haven't had time to pursue these in depth yet)
>
>Edit 2: This post might also be helpful, it has some other documentation in there (if you're doing jsonapi)

Thanks man, I'll look into that right now.

**Comment by jaydorsey** [comment]
> Does the schema have to be an exact match with the response? Like the values and such, since my values are randomly generated with Faker this might be a problem.

It shouldn't have to be, unless you've hard-coded some values (like an `enum` type with an array of acceptable values or whatever)

I think you're on to something though, because I glanced over your error message before:

    array in schema 21d9ae67-0c16-5104-a531-ae0d04a887f1

That last bit is obviously a UUID, probably from a response.

For reference, your original schema at the top level (reformatted because the tab's expansion is hard to read on github:

```
schema type: :array,
properties: {
  data: {
    type: :array,
    items: {
      '$ref': '#/definitions/ecommerce_eproducts_object'
    }
  },
  links: {
    type: :object,
    properties: {
      self: {
        type: :string,
        format: :url,
        example: "//ecommerce/v2/eproducts/?page[number]=1&page[size]=10"
      },
      first: {
        type: :string,
        format: :url,
        example: "//ecommerce/v2/eproducts/?page[number]=1&page[size]=10"
      },
      prev: {
        type: :string,
        format: :url,
        example: nil
      },
      next: {
        type: :string,
        format: :url,
        example: "//ecommerce/v2/eproducts/?page[number]=2&page[size]=10"
      },
      last: {
        type: :string,
        format: :url,
        example: "//ecommerce/v2/eproducts/?page[number]=10&page[size]=10"
      }
    },
    required: [ 'self', 'first', 'last']
  },
  meta: {
    '$ref': '#/definitions/meta_pagination_object'
  }
}
```

I _think_ you need to change that first line; `schema type: :array,` to `schema type: :object,` because your array is actually _inside_ of your data object, not outside. Can you test that and let me know if you get a different error?

**Comment by Kutomore** [comment]
So, you were right, all the issues left were due to typing, stuff that should be `:null` but written as `:integer` or `:string` and that kind of stuff.

I made a request by hand on postgres, and used [this tool](https://www.liquid-technologies.com/online-json-to-schema-converter) to convert it to a schema, popped it into notepad++ and went fixing all the types that were wrong, after that everything worked as expected.

Thanks for the help man, this issue is solved.
